wallet-new/server/Cardano/Wallet/API/V1/LegacyHandlers/Addresses.hs

```diff
@@ -51,10 +60,46 @@ newAddress NewAddress {..} = do
               >>= migrate
 
 -- | Verifies that an address is base58 decodable.
-verifyAddress :: Monad m => Text -> m (WalletResponse AddressValidity)
-verifyAddress address =
-    case decodeTextAddress address of
-        Right _ ->
-            return $ single $ AddressValidity True
-        Left _  ->
-            return $ single $ AddressValidity False
+getAddress
+    :: (MonadThrow m , V0.MonadWalletLogic ctx m)
+    => Text
+    -> m (WalletResponse WalletAddress)
+getAddress addrText = do
+    addr <- either
+        (throwM . InvalidAddressFormat)
+        pure
+        (decodeTextAddress addrText)
+
+    ss <- V0.askWalletSnapshot
+
+    let
+        addrId :: V0.CId V0.Addr
+        addrId =
+            V0.addressToCId addr
+
+        addrInfoMatchesAddr ::  V0.AddressInfo -> Bool
+        addrInfoMatchesAddr addrInfo =
+            addrId == V0.cwamId (V0.adiCWAddressMeta addrInfo)
+
+        getAddresses :: V0.CId V0.Wal -> [V0.AddressInfo]
+        getAddresses = V0State.getWAddresses ss V0.Ever
+
+        minfo :: Maybe (V0.CWalletMeta, V0.AddressInfo)
+        minfo =
+            asum
+            . map (\wid ->
+                (,) <$> V0State.getWalletMeta ss wid
+                    <*> List.find addrInfoMatchesAddr (getAddresses wid)
+                )
+            . V0.getWalletAddresses
```

Comment:
In general we are lucky as exchanges have typically 1 wallet (with lots of addresses in it) so this code is effectively "just" O(n) rather than being quadratic 😁 The crux of the problem is always the poor choice of data structures in the backend, which forces us to do expensive traversals rather than simply O(k) lookups.

Reply:
tfw no sql tables with indices for uberspeed